dace/transformation/transformation.py

```diff
@@ -107,29 +152,49 @@ def __init__(self, sdfg_id, state_id, subgraph, expr_index):
         # Serializable subgraph with node IDs as keys
         expr = self.expressions()[expr_index]
         self._subgraph = {expr.node_id(k): v for k, v in subgraph.items()}
-        self._subgraph_user = subgraph
+        self._subgraph_user = copy.copy(subgraph)
         self.expr_index = expr_index
 
+        # Ease-of-use API: Set new pattern-nodes with information about this
+        # instance.
+        for pname, pval in self._get_pattern_nodes().items():
+            # Create new pattern node from existing field
+            new_pnode = PatternNode(
+                pval.node if isinstance(pval, PatternNode) else type(pval))
+            new_pnode.match_instance = self
+
+            # Append existing values in subgraph dictionary
+            if pval in self._subgraph_user:
+                self._subgraph_user[new_pnode] = self._subgraph_user[pval]
+
+            # Override static field with the new node in this instance only
+            setattr(self, pname, new_pnode)
+
     @property
     def subgraph(self):
         return self._subgraph_user
 
-    def query_node(
-        self, sdfg: SDFG, pattern_node: Union[nd.Node, SDFGState]
-    ) -> Union[nd.Node, SDFGState]:
+    def apply_pattern(self, sdfg: SDFG) -> Union[Any, None]:
```

Comment:
Now it is worse than before :) I would say that `Any` is the worst (the same as no annotation), `Union` is better but is still a candidate for refactoring. Ideally, it would be nice to see a concrete return type.

Reply:
This is not the same function, but the diff makes it look like it is. apply_pattern always returned anything (as said in the documentation).